File: dart/high_scores.dart
import 'dart:math';

class HighScores {
  final List<int> scores;
  List<int> sorted;

  HighScores(List<int> scores)
      : scores = scores,
        sorted = [] {
    sorted = [...scores];
    sorted.sort((a, b) => b.compareTo(a));
  }

  int latest() {
    return scores.last;
  }

  int personalBest() {
    return sorted[0];
  }

  List<int> personalTopThree() {
    return sorted.sublist(0, min(3, sorted.length));
  }
}

File: lua/high-scores.lua
return function(scores)
    local sorted = {}
    table.move(scores, 1, #scores, 1, sorted)
    table.sort(sorted, function(a, b)
        return a > b
    end)
    local top3 = {}
    for i = 1, math.min(3, #sorted) do
        top3[i] = sorted[i]
    end
    return {
        scores = function(_)
            return scores
        end,
        latest = function(_)
            return scores[#scores]
        end,
        personal_best = function(_)
            return top3[1]
        end,
        personal_top_three = function(_)
            return top3
        end,
    }
end

